RuboCop 1.74 Unsafe Correctables 3/3

JoeCohen opened the pull request:
Fixes unsafe correctable offenses. Part 3 of 3. 
Broken into 3 parts for ease of review. This part includes cops linked below. 
For each putative offense:
- Accepts the fix if safe, or disables the cop if unsafe
- Makes follow-on changes to avoid new offenses. (Ex: removes `.to_s` inside string interpolation, divides long lines.)
- Updates .rubocop_todo

#### Cops in this PR
Check the linked documentation for comments about cop safety.
> [Rails/Blank](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsblank)
[Rails/Output](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsoutput)
[Rails/RakeEnvironment](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsrakeenvironment)
[Rails/RedundantActiveRecordAllMethod](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsredundantactiverecordallmethod)
[Rails/ResponseParsedBody](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsresponseparsedbody)
[Rails/RootPathnameMethods](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsrootpathnamemethods)
[Rails/SelectMap](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsselectmap)
[Style/BitwisePredicate](https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylebitwisepredicate)
[Style/CombinableLoops](https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylecombinableloops)
[Style/FileNull](https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylefilenull)[Style/SlicingWithRange](https://docs.rubocop.org/rubocop/1.74/cops_style.html#styleslicingwithrange)
[Style/HashEachMethods](https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylehasheachmethods)
[Style/SafeNavigation](https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylesafenavigation)

JoeCohen committed:
Style/CombinableLoops  check_enum_values

JoeCohen committed:
Restore inheritance from .rubocop_todo

JoeCohen committed:
update rubocop_todo

JoeCohen committed:
Style/CombinableLoops mushroom_mapper

JoeCohen committed:
Style/SlicingWithRange
 PatternSearch::Term

JoeCohen committed:
Style/SlicingWithRange location.rb

JoeCohen committed:
Style/SlicingWithRange Name::Spelling

JoeCohen committed:
Style/SlicingWithRange name_parse.rb

JoeCohen committed:
Style/SlicingWithRange comment_test

JoeCohen committed:
Style/SafeNavigation profile_controller.rb

JoeCohen committed:
Style/SafeNavigation eol_data_controller

JoeCohen committed:
Style/SafeNavigation Name::Create

JoeCohen committed:
Style/SafeNavigation Name::Spelling

JoeCohen committed:
Style/SafeNavigation naming_consensus.rb

JoeCohen committed:
Rails/RakeEnvironment lang.rake

Added `:environment` as a dependency to 4 tasks.
I don't understand the tasks.
But without this dependency, I got
```ruby
joe@iMac mushroom-observer % rake lang:login_admin
rake aborted!
NameError: uninitialized constant User (NameError)

    User.current = User.find(0)
    ^^^^
/Users/joe/mushroom-observer/lib/tasks/lang.rake:91:in `block (2 levels) in <top (required)>'
Tasks: TOP => lang:login_admin
(See full trace by running task with --trace)
```

JoeCohen committed:
Rails/ResponseParsedBody api2_controller_test

JoeCohen committed:
Rails/Output disable in language_exporter

JoeCohen committed:
Delete uncalled string method containing Rails/Output offense

JoeCohen committed:
Merge branch 'jdc-rubocop-1-74-unsafe-correctables-3' of github.com:MushroomObserver/mushroom-observer into jdc-rubocop-1-74-unsafe-correctables-3

JoeCohen committed:
Rails/RedundantActiveRecordAllMethod email.rake

JoeCohen committed:
Rails/RedundantActiveRecordAllMethod script/check_links

JoeCohen committed:
Rails/SelectMap observations_controller_test

JoeCohen committed:
Style/HashEachMethods eol_data.rb

JoeCohen committed:
Style/HashEachMethods script/slowest_pages

JoeCohen committed:
Disable Rails/Blank in image.rb

JoeCohen committed:
Rails/RootPathnameMethods monkey_patches.rb

JoeCohen committed:
Change Style/BitwisePredicate offense with clearer code spelling.rb

JoeCohen committed:
Style/FileNull queued_email_test

JoeCohen commented on app/extensions/string.rb:
```diff
@@ -739,10 +739,6 @@ def md5sum
 
   ### Misc Utilities ###
   #
-  def print_thing(thing)
```
`print_thing` isn't called anywhere.

nimmolo replied:
Seems useful (and intended) for debugging, so it wouldn't stay in the code base. But i didn't know about it.

JoeCohen commented on app/models/name/spelling.rb:
```diff
@@ -35,7 +35,7 @@ def suggest_alternate_spellings(str)
         (2..num).each do |i|
           next unless results.any?
-          next unless (i & 1).zero?
+          next unless i.even?
```
@nimmolo : Is my change okay?
RuboCop wanted me to change `(i & 1).zero?` to `i.nobits?(1)`. https://docs.rubocop.org/rubocop/1.74/cops_style.html#stylebitwisepredicate
But both seem like extra mental gymnastics.

nimmolo replied:
Thanks for asking. I don't understand what this is doing, though.

mo-nathan replied:
It's doing a bitwise and of i against the number 1 and then testing if that equals zero.  This will be the case if and only if i is even.  Why it was written the way it is I have no idea.  `git blame` current points @nimmolo , but I'm guessing it was copied from somewhere else.

JoeCohen commented on config/initializers/monkey_patches.rb:
```diff
@@ -6,7 +6,7 @@
 
 # Require all Ruby files in the core_extensions directory
-Dir[Rails.root.join("lib/core_extensions/**/*.rb")].each { |f| require f }
+Rails.root.glob("lib/core_extensions/**/*.rb").each { |f| require f }
```
@nimmolo Please double-check

JoeCohen commented on lib/tasks/lang.rake:
```diff
@@ -78,7 +78,7 @@ namespace :lang do
 
   desc "Log in user for import tasks."
-  task(:login) do
+  task(login: :environment) do
```
@mo-nathan @pellaea 
I added `:environment` as a dependency to 4 tasks.
- Question: Are these tasks ever used? 

- I don't understand the tasks.
But without this dependency, I got
```ruby
joe@iMac mushroom-observer % rake lang:login_admin
rake aborted!
NameError: uninitialized constant User (NameError)

    User.current = User.find(0)
    ^^^^
/Users/joe/mushroom-observer/lib/tasks/lang.rake:91:in `block (2 levels) in <top (required)>'
Tasks: TOP => lang:login_admin
(See full trace by running task with --trace)
```

mo-nathan replied:
I for sure don't intentionally use them.  However, they might get called indirectly.  In my view our translation stuff is kinda of wonky so I would move carefully.

JoeCohen commented:
@mo-nathan & @nimmolo:
I tagged you in comments where I had special questions.

JoeCohen committed:
Disable Rails/RakeEnvironment in lang.rake

Responds to [@mo-nathan comment](https://github.com/MushroomObserver/mushroom-observer/pull/2838/commits/fcf75b4c57ac7c6f0a3046c870236ceeda50e50f#r2011104991)

JoeCohen committed:
Fix typos in prior commit

mo-nathan commented on app/models/image.rb:
```diff
@@ -662,7 +662,10 @@ def validate_image_name
+    # disable cop; In theory upload_temp_file could be
+    # a blank non-empty string (e.g. " ").
+    # Don't fix unless...present?; it isn't broken.
+    unless upload_temp_file.present? # rubocop:disable Rails/Blank
```
`present?` is literally defined as `!blank?` (see https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/object/blank.rb).  I would go with Rubocop's recommendation here.  For " " `blank?` is true and `present?` is false.  If you want to allow filenames that include `" "` you would have to explicitly test for that.  However, that seems so edge casey that I think it's a bad idea to consider in our code.  If you know this is actually a likely name, then I'd reject it or rename it to something reasonable.

mo-nathan reviewed (approved):
Made a few minor points, but overall this looks great!

JoeCohen committed:
Restore print_thing utility

Responds to [@nimmolo comment](https://github.com/MushroomObserver/mushroom-observer/pull/2838/files?show-viewed-files=true&file-filters%5B%5D=#r2011274615)

JoeCohen committed:
Follow RuboCop Rails/Blank suggestion

Responds to [@mo-nathan comment](https://github.com/MushroomObserver/mushroom-observer/pull/2838#discussion_r2014536638)
